docs/source/torch.compiler_faq.rst

```diff
@@ -528,6 +528,149 @@ invokes an ``nn.Module``. This is because the outputs now depend on the
+  array. If this is not possible, one may often recover the original behavior
+  by cast the NumPy scalar to the relevant Python scalar (e.g., ``float``).
+- Negative strides: ``np.flip`` and slicing with a negative step return a copy.
+- Type promotion: NumPy's type promotion will change in NumPy 2.0. The new rules
+  are described in `NEP 50 <https://numpy.org/neps/nep-0050-scalar-promotion.html)>`__.
+  ``torch.compile`` implements NEP 50 rather than the current soon-to-be deprecated rules.
+- ``{tril,triu}_indices_from`` return arrays rather than lists of tuples.
+
+There are other feature for which we do not support tracing and we gracefully
+fallback to NumPy for their execution:
+
+- Non-numeric dtypes like datetimes, strings, chars, void, structured dtypes and recarrays.
+- Long dtypes like ``np.float128`` and ``np.complex256``.
+- ``ndarray`` subclasses.
+- Masked arrays.
+- Ufunc machinery like ``axes=[(n,k),(k,m)->(n,m)]``, ``np.add.reduce``, etc.
+- Fortran orders and, in general, any ``order=`` different to ``C``.
+- Sorting / ordering ``complex64/complex128`` arrays.
+- NumPy ``np.poly1d`` and ``np.polynomial``.
+- Positional ``out1, out2`` args in functions with 2 or more returns (``out=tuple`` does work).
+- ``__array_function__``, ``__array_interface__`` and ``__array_wrap__``.
+- ``ndarray.ctypes`` attribute not supported.
+
+Can I execute NumPy code on CUDA via ``torch.compile``?
+-------------------------------------------------------
+
+Yes you can! To do so, you may simply execute your code under ``torch.device("cuda")``
+
+.. code-block:: python
+
+   @torch.compile
+   def numpy_fn(X: ndarray, Y: ndarray): -> ndarray
+       # Compute the ndarray Z here
+       return Z
+
+
+   X = np.random.randn(1000, 1000)
+   Y = np.random.randn(1000, 1000)
+   with torch.device("cuda"):
+       Z = numpy_fn(X, Y)
+
+In this example, ``numpy_fn`` will be executed in CUDA. For this to be
+possible, ``torch.compile`` automatically moves ``X`` and ``Y`` from CPU
+to CUDA, and then it moves the result ``Z`` from CUDA to CPU. If we are
+executing this function several times in the same program run, we may want
+to avoid all these rather expensive memory copies. To do so, we just need
+to tweak our ``numpy_fn`` so that it accepts cuda Tensors and returns tensors:
+
+.. code-block:: python
+
+   @torch.compile
+   def numpy_fn(X: Tensor, Y: Tensor): -> Tensor
+       X = X.numpy()
+       Y = X.numpy()
+       # Compute Z here
+       Z = torch.from_numpy(Z)
+       return Z
+
+   X = torch.randn(1000, 1000, device="cuda")
+   Y = torch.randn(1000, 1000, device="cuda")
+   with torch.device("cuda"):
+       Z = numpy_fn(X, Y)
+
+By doing this, we explicitly create the tensors in CUDA memory, and we keep
+them there. Note that the original program would not run on eager mode now.
+If you want to run it in eager mode, you would need to call ``.numpy(force=True)``
+and perhaps doing ``Z = Z.cuda()`` before returning ``Z``. Of course, doing
```

Comment:
Actually, I start wondering what `.numpy(force=True)` does under `torch.compile`

Reply:
See https://github.com/pytorch/pytorch/pull/109636

Reply:
of course :-).